### ISSUE TITLE
[C#] Missing std_array.i under Lib/csharp

### ISSUE THREAD

**Issue by myd7349** [swig/swig#739]
There is not a `std_array.i` for C#.

I implemented one based on `Lib/csharp/std_vector.i`: [std_array.i](https://github.com/myd7349/Ongoing-Study/blob/master/c%23/SWIG/CPP11StdArray/CPP11StdArrayNative/std_array.i)

I am not sure whether it  will be helpful or not.
If I have done right, I may strip the useless comments and create a PR.

**Comment by wsfulton** [comment]
Sounds great! Basing it on std_vector.i is the right thing to do. You mention C# 6 and C# 4.5 being needed for IReadOnlyList. I don't see it using IReadOnlyList though, so will work with C# 2, which is the minimum version SWIG supports?

If you could convert the java runtime testcase to c# and create a Github pull request, that would be ideal. The runtime test is in Examples/test-suite/java/cpp11_li_std_array_runme.java and the input Examples/test-suite/cpp11_li_std_array.i. You'd need to change:
```c
#if defined(SWIGPYTHON) || defined(SWIGRUBY) || defined(SWIGJAVA)
```
to include SWIGCSHARP.

**Comment by myd7349** [comment]
@wsfulton Thanks for your reply!
I think `IReadOnlyList<>` would be a good base class for the C# wrapper of C++11 `std::array`. However, since it needs .NET Framework 4.5 or newer(it is too new for me), I didn't use it. 
Instead, I made the wrapper class derive from `IEnumerable` and `IEnumerable<>`, just like `IReadOnlyList<>` does.

Yes, It will work with .NET Framework 2.0.

When I was working on `Lib/csharp/std_array.i`, my SWIG version is 3.0.8. I have already tested it on SWIG 2.0.0, 3.0.0, 3.0.10. `Lib/csharp/std_array.i` works fine with these versions of SWIG too.
